CLI to upload arbitrary huge folder

Wauplin opened the pull request:
## What for?

Upload arbitrarily large folders in a single command line!


## How to use it?

**Install**
```
pip install git+https://github.com/huggingface/huggingface_hub@large-upload-cli
```

**Upload folder**

```
huggingface-cli upload-large-folder <repo-id> <local-path> --repo-type=dataset
```

Every minute a report is printed to the terminal with the current status. In addition to that, progress bars and errors are still displayed.

```
---------- 2024-04-26 16:24:25 (0:00:00) ----------
Files:   hashed 104/104 (22.5G/22.5G) | pre-uploaded: 0/42 (0.0/22.5G) | committed: 58/104 (24.9M/22.5G) | ignored: 0
Workers: hashing: 0 | get upload mode: 0 | pre-uploading: 6 | committing: 0 | waiting: 0
---------------------------------------------------
```

Run  `huggingface-cli large-upload --help` to see all options.

## PR documentation: 
- package reference (technical details) https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_2254/en/package_reference/hf_api#huggingface_hub.HfApi.large_upload
- upload guide (how to use it) https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_2254/en/guides/upload#upload-a-large-folder

## What does it do?

This CLI is intended to upload arbitrary large folders in a single command:
- process is split in 4 steps: hash, get upload mode, lfs upload, commit
- retry on error at each step
- multi-threaded: workers are managed with queues
- resumable: if the process is interrupted, you can re-run it. Only partially uploaded files are lost.
- files are hashed only once
- starts to upload files while other files are still been hashed
- commit at most 50 files at a time
- prevent concurrent commits
- prevent rate limits as much as possible
- prevent small commits
- retry on error for all steps

A `.cache/huggingface/` folder will be created at the root of your folder to keep track of the progress. Please do not modify these files manually. If you feel this folder got corrupted, please report it here, delete the `.huggingface/` entirely and then restart you command. Some intermediate steps will be lost but the upload process should be able to continue correctly.

## Known limitations

- cannot set a `path_in_repo` => always upload files at root of the folder. If you want to upload to a subfolder, you need to set the proper structure locally.
- cannot delete files on repo while uploading folder
- cannot set commit message/commit description
- cannot create PR by itself => you must first create a PR manually, then provide `revision`

These limitations are documented.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_2254). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

Wauplin commented:
Feedback so far:
- when connection is slow, better to reduce the number of workers. Should we do that automatically or just print a message? Reducing number of workers might not speed-up upload but at least less files are uploaded in parallel => less chances to loose progress in case of failed upload.
- terminal output is too verbose. Might be good to disable individual progress bars?
- terminal output is awful in a jupyter notebook => how can we make that more friendly (printing a report every 1 minute ends up with very long logs)
- a CTRL-C (or at most 2 CTRL+C) must stop the process. It's not the case at moment due to all try/except.

julien-c commented:
ah i rather agree with @davanstrien here

wanng-ide commented:
Can the parameters of "large-upload" be aligned to the "upload"?
huggingface-cli large-upload [repo_id] [local_path]

Wauplin commented:
@wanng-ide Agree we should aim for consistency yes. What parameters/options you would specifically change?

So far we have:
```sh
$ huggingface-cli large-upload --help
usage: huggingface-cli <command> [<args>] large-upload [-h] [--repo-type {model,dataset,space}]
                                                       [--revision REVISION] [--private]
                                                       [--include [INCLUDE ...]] [--exclude [EXCLUDE ...]]
                                                       [--token TOKEN] [--num-workers NUM_WORKERS]
                                                       repo_id local_path
```

```sh
$ huggingface-cli upload --help 
usage: huggingface-cli <command> [<args>] upload [-h] [--repo-type {model,dataset,space}]
                                                 [--revision REVISION] [--private] [--include [INCLUDE ...]]
                                                 [--exclude [EXCLUDE ...]] [--delete [DELETE ...]]
                                                 [--commit-message COMMIT_MESSAGE]
                                                 [--commit-description COMMIT_DESCRIPTION] [--create-pr]
                                                 [--every EVERY] [--token TOKEN] [--quiet]
                                                 repo_id [local_path] [path_in_repo]
```

wanng-ide commented:
> @wanng-ide Agree we should aim for consistency yes. What parameters/options you would specifically change?
> 
> So far we have:
> 
> ```shell
> $ huggingface-cli large-upload --help
> usage: huggingface-cli <command> [<args>] large-upload [-h] [--repo-type {model,dataset,space}]
>                                                        [--revision REVISION] [--private]
>                                                        [--include [INCLUDE ...]] [--exclude [EXCLUDE ...]]
>                                                        [--token TOKEN] [--num-workers NUM_WORKERS]
>                                                        repo_id local_path
> ```
> 
> ```shell
> $ huggingface-cli upload --help 
> usage: huggingface-cli <command> [<args>] upload [-h] [--repo-type {model,dataset,space}]
>                                                  [--revision REVISION] [--private] [--include [INCLUDE ...]]
>                                                  [--exclude [EXCLUDE ...]] [--delete [DELETE ...]]
>                                                  [--commit-message COMMIT_MESSAGE]
>                                                  [--commit-description COMMIT_DESCRIPTION] [--create-pr]
>                                                  [--every EVERY] [--token TOKEN] [--quiet]
>                                                  repo_id [local_path] [path_in_repo]
> ```

what about: huggingface-cli large-upload [local_path] [path_in_repo]
ADD [path_in_repo]

Wauplin commented:
I'm not sure to understand what's the purpose of the `ADD` keyword

rom1504 commented:
Will this only be a clip or also a python function? I liked the python API for upload folder. Convenient to automate sending many datasets in python rather than bash

Wauplin commented:
> Will this only be a cli or also a python function?

Yes, that's the goal. At the moment, it is defined as a standalone method `large_upload()` (see [here](https://github.com/huggingface/huggingface_hub/pull/2254/files#diff-fabf05bef8228544f4baec532db8d2513ebbcd308aa9edeb4f3a5e0d554e8c58R30)). In a final version, we will probably add it to `HfApi` client.

rom1504 commented:
I'm using it to upload a few 300GB datasets. The standard upload function was taking more than 30min just to hash the files and then was crashing half way. This seems to be working much better.

rom1504 commented:
Ok I got one more piece of feedback actually...
Looks like this tool is *too* fast :)

It seems to be killing my box for a few hours (after uploading at 80MB/s for a few hours). I don't really get how that's possible yet.

What would you advise to reduce the speed a bit / reduce the number of simultaneous connections ?

Wauplin commented:
Wow, this is an unexpected problem :smile: I can think of two ways of reducing the upload speed:
1. don't use `hf_transfer` if you were previously doing it
2. set `--num-workers=1` (or 2/3) to reduce the number of workers uploading files in parallel. However there is currently no way to throttle the connection from `huggingface_hub` directly. There is a separate issue for that (see https://github.com/huggingface/huggingface_hub/issues/2118#issuecomment-2157587132) but I don't think we'll ever work on such a feature. You can set this from your setup with a proxy I believe, though it's quite hacky.

Wauplin commented:
@ducha-aiki Which Python version are you using? Could you try to upgrade to 3.10 and let me know if it still happens? I suspect `queue.Queue[JOB_ITEM_T]` to be forbidden in Python 3.8

ducha-aiki commented:
Oh no, my favorite 2 yo environment... you got me, that was Python 3.8.12. 
Trying on 3.10 now, no error, but also nothing got uploaded...the repo is either not created( tried without repo), or empty (when tried to creating repo first)

```
INFO:huggingface_hub.large_upload:

##########
Large upload status:
  Progress:
    57/57 hashed files (15.5M/15.5M)
    57/57 preuploaded LFS files (15.5M/15.5M)
    57/57 committed files (15.5M/15.5M)
    (0 gitignored files)
  Jobs:
    sha256: 0 workers (0 items in queue)
    get_upload_mode: 0 workers (0 items in queue)
    preupload_lfs: 0 workers (0 items in queue)
    commit: 0 workers (0 items in queue)
  Elapsed time: 0:01:00
  Current time: 2024-06-10 12:44:23
##########

```

The folder structure I am trying to upload:

`data/*.parquet`

ducha-aiki commented:
@Wauplin update: the files finally appeared now, although they pretend to be added 6 min ago. Anything, everything works now, thank you :)

<img width="1175" alt="image" src="https://github.com/huggingface/huggingface_hub/assets/4803565/97260d4e-4b10-40f3-aeb7-6eef8cc00543">

Wauplin commented:
Thanks to everyone who've tested this feature! :heart: Feedback has been very valuable to improve user experience. I think that I've addressed most of it. I have added some documentation + cleaned a bit the implementation. PR is ready to be reviewed cc @LysandreJik @osanseviero 

Test-wise I did not do much. What I can do is to run a large upload and check that all files have been uploaded. But I'm a bit lazy to unittest each part one by one. Let me know what you think :see_no_evil:  

Let's finally ship this! :rocket: 

Note: oh and last thing, I don't like the name `large_upload`. Any suggestion?

LysandreJik commented on docs/source/en/guides/upload.md:
```diff
@@ -103,6 +103,70 @@ set, files are uploaded at the root of the repo.
+
+<Tip>
+
+For large uploads, you have to set `repo_type="model"` or `--repo-type=model` explicitly. Usually this information is implicit in all other `HfApi` methods. This is to avoid having data uploaded to a repository with a wrong type. If that's the case, you'll have to re-upload everything.
```
Not a bad idea

osanseviero replied:
```suggestion
For large uploads, you have to set `repo_type="model"` or `--repo-type=model` explicitly. Usually, this information is implicit in all other `HfApi` methods. This is to avoid having data uploaded to a repository with a wrong type. If that's the case, you'll have to re-upload everything.
```

LysandreJik commented on src/huggingface_hub/hf_api.py:
```diff
@@ -5149,6 +5150,123 @@ def delete_folder(
+
+        <Tip warning={true}>
+
+        While being much more robust to upload large folders, `large_upload` is more limited than [`upload_folder`] feature-wise. In practice:
+            - you cannot set a custom `path_in_repo`. If you want to upload to a subfolder, you need to set the proper structure locally.
+            - you cannot set a custom `commit_message` and `commit_description` since multiple commits are created.
+            - you cannot delete from the repo while uploading. Please make a separate commit first.
+            - you cannot create a PR directly. Please create a PR first and then commit to it by passing `revision`.
```
These limitations are important, what about putting them in the `upload.md` additionally?

Wauplin replied:
addressed in https://github.com/huggingface/huggingface_hub/pull/2254/commits/602f299a399df31b4608b9d45af1b310728bb650

LysandreJik commented on src/huggingface_hub/hf_api.py:
```diff
@@ -5149,6 +5150,123 @@ def delete_folder(
+
+        `large_upload` process is as follow:
+            1. (Check parameters and setup.)
+            2. Create repo is missing.
```
```suggestion
            2. Create repo if missing.
```

LysandreJik commented on docs/source/en/guides/upload.md:
```diff
@@ -103,6 +103,70 @@ set, files are uploaded at the root of the repo.
+
+- **Start small**: We recommend starting with a small amount of data to test your upload script. It's easier to iterate on a script when failing takes only a little time.
+- **Expect failures**: Streaming large amounts of data is challenging. You don't know what can happen, but it's always best to consider that something will fail at least once -no matter if it's due to your machine, your connection, or our servers. For example, if you plan to upload a large number of files, it's best to keep track locally of which files you already uploaded before uploading the ne ...
+- **Use `hf_transfer`**: this is a Rust-based [library](https://github.com/huggingface/hf_transfer) meant to speed up uploads on machines with very high bandwidth. To use `hf_transfer`:
```
Is `hf_transfer` still recommended when using the large upload feature?

Wauplin replied:
Yes it is! In that case, the number of workers pre-uploading files is limited to 1 since `hf_transfer` maxes out the CPU. This means the files will be hashed in parallel and then uploaded 1 by 1 which is a good behavior IMO.

LysandreJik reviewed (approved):
Great PR @Wauplin! I've used it extensively for the 3.1 release and it was working flawlessly.

I agree that forcing the user to make choices ahead of time like the repo type makes sense given the size of the files that will be uploaded.

osanseviero commented on docs/source/en/guides/upload.md:
````diff
@@ -103,6 +103,70 @@ set, files are uploaded at the root of the repo.
+---------------------------------------------------
+```
+
+First, the repo is created if it didn't exist before. Then the local folder is scanned for files to upload. For each file, we try to recover metadata information (from a previously interrupted upload). From there, it is able to launch workers and print an update status every 1 minute. Here we can see that 5 files have already been hashed but not pre-uploaded. 5 workers are pre-uploading files whi ...
````
```suggestion
First, the repo is created if it didn't exist before. Then, the local folder is scanned for files to upload. For each file, we try to recover metadata information (from a previously interrupted upload). From there, it is able to launch workers and print an update status every 1 minute. Here, we can see that 5 files have already been hashed but not pre-uploaded. 5 workers are pre-uploading files while the 11 others are waiting for a task.
```

osanseviero commented on src/huggingface_hub/_large_upload.py:
```diff
@@ -0,0 +1,567 @@
+    repo_id = repo_url.repo_id
+
+    # 3. List files to upload
+    #    Taken from '_prepare_upload_folder_additions(...)'
```
Does this still exist? Or should it refer to `push_to_hub` instead?

osanseviero replied:
Or maybe not refer to any other file if this is different

osanseviero commented on src/huggingface_hub/_large_upload.py:
```diff
@@ -0,0 +1,567 @@
+
+    # 3. List files to upload
+    #    Taken from '_prepare_upload_folder_additions(...)'
+    paths_list = [
+        get_local_upload_paths(folder_path, relpath)
+        for relpath in filter_repo_objects(
+            (path.relative_to(folder_path).as_posix() for path in folder_path.glob("**/*") if path.is_file()),
+            allow_patterns=allow_patterns,
+            ignore_patterns=ignore_patterns,
+        )
+    ]
+    logger.info(f"Found {len(paths_list)} candidate files to upload")
```
Consider splitting into two steps for better readability (`filtered_paths` and then `paths_list`)

Wauplin replied:
addressed in https://github.com/huggingface/huggingface_hub/pull/2254/commits/86edb11ae90400aeeafb195cd78808454069fbe4

osanseviero commented on src/huggingface_hub/_large_upload.py:
```diff
@@ -0,0 +1,567 @@
+                logger.debug("Job: get upload mode (no other worker getting upload mode)")
+
+            # 7. Preupload LFS file if at least 1 file
+            #    Skip if hf_transfer is enabled and there is already a worker preuploading LFS
+            elif status.queue_preupload_lfs.qsize() > 0 and (
+                status.nb_workers_preupload_lfs == 0 or not constants.HF_HUB_ENABLE_HF_TRANSFER
+            ):
+                status.nb_workers_preupload_lfs += 1
+                next_job = (WorkerJob.PREUPLOAD_LFS, _get_one(status.queue_preupload_lfs))
+                logger.debug("Job: preupload LFS")
+
+            # 8. Compute sha256 if at least 1 file
+            elif status.queue_sha256.qsize() > 0:
+                status.nb_workers_sha256 += 1
+                next_job = (WorkerJob.SHA256, _get_one(status.queue_sha256))
+                logger.debug("Job: sha256")
+
+            # 9. Get upload mode if at least 1 file
+            elif status.queue_get_upload_mode.qsize() > 0:
+                status.nb_workers_get_upload_mode += 1
+                next_job = (WorkerJob.GET_UPLOAD_MODE, _get_n(status.queue_get_upload_mode, 50))
+                logger.debug("Job: get upload mode")
+
+            # 10. Commit if at least 1 file
+            elif status.nb_workers_commit == 0 and status.queue_commit.qsize() > 0:
+                status.nb_workers_commit += 1
+                next_job = (WorkerJob.COMMIT, _get_n(status.queue_commit, 25))
+                logger.debug("Job: commit")
+
+            # 11. If all queues are empty, exit
+            elif all(metadata.is_committed or metadata.should_ignore for _, metadata in status.items):
+                logger.info("All files have been processed! Exiting worker.")
+                return
+
+            # 12. If no task is available, wait
+            else:
+                status.nb_workers_waiting += 1
+                next_job = (WorkerJob.WAIT, [])
+                logger.debug(f"No task available, waiting... ({WAITING_TIME_IF_NO_TASKS}s)")
+
```
Nit: maybe move this to a function that returns the next job

osanseviero reviewed (commented):
Looking nice! I did a quick first pass, but this is very cool :fire:

osanseviero commented:
I wonder if we should rename to `upload_large_folder`

Wauplin commented:
> I wonder if we should rename to upload_large_folder

Good idea! Renamed in https://github.com/huggingface/huggingface_hub/pull/2254/commits/0dc048ac85cc60b8b4082851b9d1bf010b3e768e

ArthurConmy commented:
![image](https://github.com/user-attachments/assets/20bfdcb6-f5d5-46a0-b7de-0e6514cfedc0)

The code snippet in the upload guide has a typo. It should say `api.upload_large_folder` not `api.upload_folder`

Wauplin commented:
Thanks for noticing and reporting @ArthurConmy! I just fixed it in https://github.com/huggingface/huggingface_hub/pull/2254/commits/ea765be18a1089d57a474e5eab59b4e7789e6a7c

osanseviero commented on docs/source/en/guides/upload.md:
```diff
@@ -103,6 +103,80 @@ set, files are uploaded at the root of the repo.
+
+While being much more robust to upload large folders, `upload_large_folder` is more limited than [`upload_folder`] feature-wise. In practice:
+- you cannot set a custom `path_in_repo`. If you want to upload to a subfolder, you need to set the proper structure locally.
+- you cannot set a custom `commit_message` and `commit_description` since multiple commits are created.
```
For a follow-up PR: I think we could support this. The PR title could be the commit message (and then just add "Part 1" or something like that)

osanseviero commented on docs/source/en/guides/upload.md:
```diff
@@ -103,6 +103,80 @@ set, files are uploaded at the root of the repo.
+- you cannot set a custom `path_in_repo`. If you want to upload to a subfolder, you need to set the proper structure locally.
+- you cannot set a custom `commit_message` and `commit_description` since multiple commits are created.
+- you cannot delete from the repo while uploading. Please make a separate commit first.
+- you cannot create a PR directly. Please create a PR first and then commit to it by passing `revision`.
```
Why?

Wauplin replied:
Because we would have to ensure some logic to ensure the script is resumable. If `create_pr=True` is passed, we would have to use `create_pull_request` and then commit files to it. But if the process is interrupted and the user relaunch it, it would have to automatically resume to the last PR. I didn't want to implement this logic in the first PR as it's quite orthogonal to the upload logic. Since large uploads are meant for long lasting processes, I don't think that manually creating PRs will be a blocker here.

Worth case, the user can create a script to call `create_pull_request` and then `large_upload`. I have added a comment in the docs for that (see https://github.com/huggingface/huggingface_hub/pull/2254/commits/ba6e4a87ef6066d8d9028bcecaf0490d91337ce3).

osanseviero commented on docs/source/en/guides/upload.md:
```diff
@@ -103,6 +103,80 @@ set, files are uploaded at the root of the repo.
+
+There are some limitations to be aware of when dealing with a large amount of data in your repo. Given the time it takes to stream the data, getting an upload/push to fail at the end of the process or encountering a degraded experience, be it on hf.co or when working locally, can be very annoying.
+
+Check out our [Repository limitations and recommendations](https://huggingface.co/docs/hub/repositories-recommendations) guide for best practices on how to structure your repositories on the Hub. Next, let's move on with some practical tips to make your upload process as smooth as possible.
```
```suggestion
Check out our [Repository limitations and recommendations](https://huggingface.co/docs/hub/repositories-recommendations) guide for best practices on how to structure your repositories on the Hub. Let's move on with some practical tips to make your upload process as smooth as possible.
```

osanseviero commented on docs/source/en/guides/upload.md:
```diff
@@ -103,6 +103,80 @@ set, files are uploaded at the root of the repo.
+- **Expect failures**: Streaming large amounts of data is challenging. You don't know what can happen, but it's always best to consider that something will fail at least once -no matter if it's due to your machine, your connection, or our servers. For example, if you plan to upload a large number of files, it's best to keep track locally of which files you already uploaded before uploading the ne ...
+- **Use `hf_transfer`**: this is a Rust-based [library](https://github.com/huggingface/hf_transfer) meant to speed up uploads on machines with very high bandwidth. To use `hf_transfer`:
+    1. Specify the `hf_transfer` extra when installing `huggingface_hub`
+       (e.g. `pip install huggingface_hub[hf_transfer]`).
```
```suggestion
       (i.e., `pip install huggingface_hub[hf_transfer]`).
```

osanseviero reviewed (approved):
Cool stuff :fire:

FurkanGozukara commented:
how do i give repo type this command fails and you didn't write in first post :)

```
!huggingface-cli upload-large-folder "MonsterMMORPG/FLUX_Kohya_SS_Massive_Research_Part5" "/home/Ubuntu/apps/StableSwarmUI/Models/Lora"

print(".\n.\nUPLOAD COMPLETED")
```

Wauplin commented:
@FurkanGozukara you an use `huggingface-cli upload-large-folder --help` to learn how to use the CLI.
To pass a repo type, you must add `--repo-type=dataset` for instance.

FurkanGozukara commented:
it prints too many messages too frequently while uploading it even crashed my notebook

can we limit it to display status at the same line or maybe lesser more compact?

it prints like 100 messages like this every second

![image](https://github.com/user-attachments/assets/0b74c4c6-083c-42f3-9c8b-db7705c1b760)

FurkanGozukara commented:
it says uploads completed run triple times but there are no files in repo


```
You are about to upload a large folder to the Hub using `huggingface-cli upload-large-folder`. This is a new feature so feedback is very welcome!

A few things to keep in mind:
  - Repository limits still apply: https://huggingface.co/docs/hub/repositories-recommendations
  - Do not start several processes in parallel.
  - You can interrupt and resume the process at any time. The script will pick up where it left off except for partially uploaded files that would have to be entirely reuploaded.
  - Do not upload the same folder to several repositories. If you need to do so, you must delete the `./.cache/huggingface/` folder first.

Some temporary metadata will be stored under `/home/Ubuntu/apps/StableSwarmUI/Models/Lora/.cache/huggingface`.
  - You must not modify those files manually.
  - You must not delete the `./.cache/huggingface/` folder while a process is running.
  - You can delete the `./.cache/huggingface/` folder to reinitialize the upload state when process is not running. Files will have to be hashed and preuploaded again, except for already committed files.

For more details, run `huggingface-cli upload-large-folder --help` or check the documentation at https://huggingface.co/docs/huggingface_hub/guides/upload#upload-a-large-folder.
Repo created: https://huggingface.co/datasets/MonsterMMORPG/FLUX_Kohya_SS_Massive_Research_Part5
Found 181 candidate files to upload
Recovering from metadata files: 100%|████████| 181/181 [00:00<00:00, 564.82it/s]
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.
All files have been processed! Exiting worker.



---------- 2024-08-28 17:53:59 (0:00:00) ----------
Files:   hashed 181/181 (371.5G/371.5G) | pre-uploaded: 161/161 (371.5G/371.5G) | committed: 181/181 (371.5G/371.5G) | ignored: 0
Workers: hashing: 0 | get upload mode: 0 | pre-uploading: 0 | committing: 0 | waiting: 0
---------------------------------------------------

---------- 2024-08-28 17:54:00 (0:00:01) ----------
Files:   hashed 181/181 (371.5G/371.5G) | pre-uploaded: 161/161 (371.5G/371.5G) | committed: 181/181 (371.5G/371.5G) | ignored: 0
Workers: hashing: 0 | get upload mode: 0 | pre-uploading: 0 | committing: 0 | waiting: 0
---------------------------------------------------
INFO:huggingface_hub._upload_large_folder:
---------- 2024-08-28 17:54:00 (0:00:01) ----------
Files:   hashed 181/181 (371.5G/371.5G) | pre-uploaded: 161/161 (371.5G/371.5G) | committed: 181/181 (371.5G/371.5G) | ignored: 0
Workers: hashing: 0 | get upload mode: 0 | pre-uploading: 0 | committing: 0 | waiting: 0
---------------------------------------------------
.
.
UPLOAD COMPLETED
```

Wauplin commented on src/huggingface_hub/_upload_large_folder.py:
```diff
@@ -0,0 +1,572 @@
+    # 1. Check args and setup
+    if repo_type is None:
+        raise ValueError(
+            "For large uploads, `repo_type` is explicitly required. Please set it to `model`, `dataset` or `space`."
```
addressed in https://github.com/huggingface/huggingface_hub/pull/2254/commits/cdfb27fff11f2a286fff3042715a828bb531103e

Wauplin commented:
> can we limit it to display status at the same line or maybe lesser more compact?

@FurkanGozukara yes you can do that by passing `--no-bars` and `--no-reports` in the command line. I have added a command to show it more prominently.

Wauplin commented:
> it says uploads completed run triple times but there are no files in repo

Have you tried to reupload the same folder to multiple locations? If yes, only the first upload will be correct. As mentioned in the little help section:

> - You can delete the `./.cache/huggingface/` folder to reinitialize the upload state when process is not running. Files will have to be hashed and preuploaded again, except for already committed files.

I suspect that your local metadata says the files are already uploaded. You can delete it and rerun the command.

Wauplin commented:
Thanks everyone involved in this PR! Feedback from everyone has been immensely valuable to shape this feature. I hope it'll now benefit to as much users as possible! :heart_hands: 

Time to merge!

FurkanGozukara commented:
> > it says uploads completed run triple times but there are no files in repo
> 
> Have you tried to reupload the same folder to multiple locations? If yes, only the first upload will be correct. As mentioned in the little help section:
> 
> > * You can delete the `./.cache/huggingface/` folder to reinitialize the upload state when process is not running. Files will have to be hashed and preuploaded again, except for already committed files.
> 
> I suspect that your local metadata says the files are already uploaded. You can delete it and rerun the command.

it appeared later for some reason

FurkanGozukara commented:
@Wauplin when do files appear on repo exactly?

upload completed like 2 min and it has been like 20 min files still not there and no info anywhere

e.g. folder : https://huggingface.co/MonsterMMORPG/video_tutorial_mass_compute/tree/main

FurkanGozukara commented:
@Wauplin can we set sub folder path right now

 I tried like this for subfolder and it failed

```

!huggingface-cli upload-large-folder "MonsterMMORPG/3D-Cartoon-Style-FLUX" r"C:\flux training\upload" --repo-type=model --no-bars

print(".\n.\nUPLOAD COMPLETED")
```

![image](https://github.com/user-attachments/assets/d5362fd5-bf33-47bf-99e2-db3ecf9b5364)


![image](https://github.com/user-attachments/assets/3762bdf2-fdcb-4cff-b940-52f79238d0b0)

Wauplin commented:
Hi @FurkanGozukara, no this is currently not possible. See known limitations in the [PR description](https://github.com/huggingface/huggingface_hub/pull/2254#issue-2265949919):

> 
cannot set a path_in_repo => always upload files at root of the folder. If you want to upload to a subfolder, you need to set the proper structure locally.

FurkanGozukara commented:
> Hi @FurkanGozukara, no this is currently not possible. See known limitations in the [PR description](https://github.com/huggingface/huggingface_hub/pull/2254#issue-2265949919):
> 
> > cannot set a path_in_repo => always upload files at root of the folder. If you want to upload to a subfolder, you need to set the proper structure locally.

I saw it ty. How do I set proper structure locally? I tried like above screenshot it failed :/

Wauplin commented:
I don't understand the error to be honest. The message says the provided path must be a folder so I guess it's a problem with the input parameters. Could you try to provide it with a python script instead of calling it from the CLI? That would help narrow down the problem.

```py
from huggingface_hub import HfApi

api = HfApi()
api.upload_large_folder(repo_id=..., repo_type=..., ...)
```

More info in https://huggingface.co/docs/huggingface_hub/main/en/package_reference/hf_api#huggingface_hub.HfApi.upload_large_folder

FurkanGozukara commented:
> ```python
> from huggingface_hub import HfApi
> 
> api = HfApi()
> api.upload_large_folder(repo_id=..., repo_type=..., ...)
> ```

it both gave error and worked weird

i see files uploaded

![image](https://github.com/user-attachments/assets/3c66c65d-0aa3-494e-b577-7cb780379369)

Wauplin commented:
This errors happens only if token is not valid. Doesn't seem related to tool itself.
In any case, glad to know the files have been uploaded :)

FurkanGozukara commented:
> This errors happens only if token is not valid. Doesn't seem related to tool itself. In any case, glad to know the files have been uploaded :)

thank you so much this new upload is amazing

FurkanGozukara commented:
@Wauplin  the new upload works amazing

but i get this warning / error - lots of times - uploading around 70 gb

is this expected? files are uploaded to repo successfully 

![image](https://github.com/user-attachments/assets/954fb3bc-2a34-4bf2-8027-a9f84405319b)

Wauplin commented:
This is not expected no but I supposed it has to do with how jupyter notebooks handle logs. Nothing much to worry about.

FurkanGozukara commented:
@Wauplin i have been rate-limited for the first time

could this be related to new method?

None of the models uploaded fully, i trust resume capability at the moment :D

![image](https://github.com/user-attachments/assets/99d4be64-6f92-4f09-afbb-8531bc61e656)

![image](https://github.com/user-attachments/assets/ce31ce22-a285-4240-b502-9834a63def85)

FurkanGozukara commented:
@Wauplin i waited several hours

restarted process, and definitely it is hitting api limit when verifying which files were accurately uploaded

i don't know if can be solved or not . i have lots of small files

just letting you know


![image](https://github.com/user-attachments/assets/f0b3267c-90ab-4810-afd3-af021b6249b6)

Wauplin commented:
Hi @FurkanGozukara sorry for the inconvenience. How many files are we talking about and which size for each of them? And which file extension ? Also, are the files uploaded as regular or LFS files? This info would help knowing use cases that are not handled perfectly

FurkanGozukara commented:
> Hi @FurkanGozukara sorry for the inconvenience. How many files are we talking about and which size for each of them? And which file extension ? Also, are the files uploaded as regular or LFS files? This info would help knowing use cases that are not handled perfectly

10581 files

Around 44 files are big like 6 7 gb rest are images like 1 - 2 mb

FurkanGozukara commented:
@Wauplin 
here full list



```
Extension: 
  Total files: 1
  Average size: 4.00 KB
  Min size: 4.00 KB
  Max size: 4.00 KB

Extension: .metadata
  Total files: 10581
  Average size: 4.00 KB
  Min size: 4.00 KB
  Max size: 4.00 KB

Extension: .safetensors
  Total files: 44
  Average size: 6.35 GB
  Min size: 3.97 GB
  Max size: 6.46 GB

Extension: .json
  Total files: 3
  Average size: 4.00 KB
  Min size: 4.00 KB
  Max size: 4.00 KB

Extension: .toml
  Total files: 6
  Average size: 4.00 KB
  Min size: 4.00 KB
  Max size: 4.00 KB

Extension: .npz
  Total files: 5240
  Average size: 260.00 KB
  Min size: 260.00 KB
  Max size: 260.00 KB

Extension: .png
  Total files: 44
  Average size: 1.48 MB
  Min size: 1.29 MB
  Max size: 1.64 MB

Extension: .txt
  Total files: 42
  Average size: 4.00 KB
  Min size: 4.00 KB
  Max size: 4.00 KB

Extension: .jpg
  Total files: 5200
  Average size: 640.86 KB
  Min size: 220.00 KB
  Max size: 1.27 MB
```

Wauplin commented:
Thanks for the details! I don't have the bandwidth to check that now but it can definitely prove useful at some point. Can I ask you to open a new issue dedicated to it? Describing your problem when getting rate limited with this structure of repo. Thanks in advance!
